test/rules-ceddl-fullstory.spec.ts

```diff
@@ -1,149 +1,159 @@
-
-    expectGlobal('digitalData').cart.item.push(secondProduct);
-
-    [eventName, payload] = expectFS('event');
-    expectEqual(eventName, 'cart_item');
-    expectEqual(payload.productInfo.sku, secondProduct.productInfo.sku);
-    expectUndefined(payload, 'linkedProduct');
-  });
-
-  it('it should not send CEDDL cart item products to FS.event on load', () => {
-    expectGlobal('digitalData').cart.item.push(basicDigitalData.product[0]);
-
-    ExpectObserver.getInstance().create({
-      rules: [expectRule('fs-event-ceddl-cart-item')],
-    });
-
-    const fs = expectGlobal('FS');
-    expectNoCalls(fs, 'event');
-  });
-
-  it('it should send CEDDL page properties to FS.event', () => {
-    expectGlobal('digitalData').page.framework = 'react'; // inject custom property
-
-    ExpectObserver.getInstance().create({
-      rules: [expectRule('fs-event-ceddl-page')], readOnLoad: true,
+  getRulesetTestEnvironments().forEach((testEnv) => {
+    describe(`test environment: ${testEnv.name}`, () => {
+      let testHarness: RulesetTestHarness;
+
+      beforeEach(async () => {
+        testHarness = await testEnv.createTestHarness(ceddlRules, { digitalData: basicDigitalData });
+      });
+
+      afterEach(async () => {
+        await testHarness.tearDown();
+      });
+
+      after(async () => {
+        await testEnv.tearDown();
+      });
+
+      it('sends the first CEDDL product to FS.event', async () => {
+        await testHarness.execute(() => {
+          (globalThis as any).digitalData.product[0].attributes = { customProp: 'foo' };
+        });
+
+        const [eventName, payload] = await testHarness.popEvent();
+        expectEqual(eventName, 'product');
+        expectMatch(payload, basicDigitalData.product[0].productInfo, 'sku', 'productID', 'productName');
+        expectMatch(payload, basicDigitalData.product[0].category, 'primaryCategory');
+        expectEqual(payload.customProp, 'foo');
+      });
+
+      it('sends CEDDL cart to FS.event', async () => {
+        await testHarness.execute(() => {
+          (globalThis as any).digitalData.cart.attributes = { promotion: 'LaborDay2020' };
+        });
+
+        const [eventName, payload] = await testHarness.popEvent();
+        expectEqual(eventName, 'cart');
+        expectEqual(payload.cartID, basicDigitalData.cart.cartID);
+        expectEqual(payload.basePrice, basicDigitalData.cart.price.basePrice);
+        expectEqual(payload.promotion, 'LaborDay2020');
+      });
+
+      it('sends dynamic CEDDL cart item additions to FS.event', async () => {
+        const firstProduct = basicDigitalData.product[0];
+
+        const secondProduct = {
+          ...firstProduct,
+          productInfo: {
+            ...firstProduct.productInfo,
+            sku: 'test',
+          },
+        };
+
+        await testHarness.execute(([localFirstProduct]) => {
+          (globalThis as any).digitalData.cart.item.push(localFirstProduct);
+        }, [firstProduct]);
```

Comment:
These invocations may look a little odd, but they match Playwright's `page.evaluate` calls. Whereas `node` is able to capture variables defined in the outer scope, Playwright callbacks for JavaScript evaluation in the browser require you to explicitly pass in args from the outer scope as an array. Here, we're explicitly passing in `firstProduct` from the outer scope so it can be referenced as `localFirstProduct` in the Playwright-managed browser environment.